Add rolling-file appender to core logging

pgayvallet opened the pull request:
## Summary

Fix https://github.com/elastic/kibana/issues/56291

This PR adds a new `rolling-file` appender to core's logging system.

The appender currently supports two triggering policy: time-based (`time-interval`) and size-base (`size-limit`), and one rolling strategy: appending a numeric-based suffix at the end of the file when rolling it (`numeric`). See `src/core/server/logging/README.md` in the PR for more documentation

### Examples

#### Rolling every 24h

```yaml
logging:
  appenders:
    lets-roll:
      kind: rolling-file
      path: /var/logs/kibana.log
      policy:
        kind: time-interval
        interval: 24h
        modulate: true
      strategy:
        kind: numeric
        pattern: '-%i'
        max: 2
      layout:
        kind: pattern
```

This appender is configured to roll the file every 24h hours (`interval: 24h`) everyday at 0:00am (`modulate: true`). it will retain a maximum of 2 (`max: 2`) rolled files:

- During the first rollover, kibana.log is renamed to kibana-1.log. A new kibana.log file is created and starts
  being written to.
- During the second rollover, kibana-1.log is renamed to kibana-2.log and kibana.log is renamed to kibana-1.log.
  A new kibana.log file is created and starts being written to.
- During the third and subsequent rollovers, kibana-2.log is deleted, kibana-1.log is renamed to kibana-2.log and
  kibana.log is renamed to kibana-1.log. A new kibana.log file is created and starts being written to.

#### Rolling once the file reaches 50mb

```yaml
logging:
  appenders:
    rock-n-roll:
      kind: rolling-file
      path: /var/logs/kibana.log
      policy:
        kind: size-limit
        size: 50mb
      strategy:
        kind: numeric
        pattern: '-%i'
        max: 2
      layout:
        kind: pattern
```

This appender is configured to roll the file once it reaches `50mb`. The same rolling strategy than previous example applies. 

### Checklist

- [x] [Documentation](https://www.elastic.co/guide/en/kibana/master/development-documentation.html) was added for features that require explanation or tutorials
- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios

kibanamachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91290/)
* Commit: 1ae02893c04952a64641ca40a79a30abecdcd77e
* This comment will update when the build is complete

### Failed CI Steps
* [Setup Build Environment and Dependencies](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91290/execution/node/106/log)
* [Setup Build Environment and Dependencies](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91290/execution/node/112/log)

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91290/","number":"91290","commit":"1ae02893c04952a64641ca40a79a30abecdcd77e"}]}
PIPELINE-->

kibanamachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91319/)
* Commit: b5f098ffb4cbd1828adb4d4a96a637b2d32f70c0
* This comment will update when the build is complete

### Failed CI Steps
* [Check Doc API Changes](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91319/execution/node/371/log)

### History
* :broken_heart: [Build #91290](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91290/) failed 1ae02893c04952a64641ca40a79a30abecdcd77e

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91290/","number":"91290","commit":"1ae02893c04952a64641ca40a79a30abecdcd77e"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91319/","number":"91319","commit":"b5f098ffb4cbd1828adb4d4a96a637b2d32f70c0"}]}
PIPELINE-->

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91326/)
* Commit: a56c61e9ff5f3e4172ca589d15a196152c4838ec

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Distributable file count
| id | [before](https://github.com/elastic/kibana/commit/67564b9776abd608434c95375968598bfa082849) | [after](https://github.com/elastic/kibana/commit/a56c61e9ff5f3e4172ca589d15a196152c4838ec) | diff |
| --- | --- | --- | --- |
| `default` | 43182 | 43199 | +17 |
| `oss` | 22579 | 22596 | +17 |

### History
* :broken_heart: [Build #91319](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91319/) failed b5f098ffb4cbd1828adb4d4a96a637b2d32f70c0
* :broken_heart: [Build #91290](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91290/) failed 1ae02893c04952a64641ca40a79a30abecdcd77e

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91290/","number":"91290","commit":"1ae02893c04952a64641ca40a79a30abecdcd77e"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91319/","number":"91319","commit":"b5f098ffb4cbd1828adb4d4a96a637b2d32f70c0"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91326/","number":"91326","commit":"a56c61e9ff5f3e4172ca589d15a196152c4838ec"}]}
PIPELINE-->

kibanamachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91399/)
* Commit: 6b6fc49779127fcd80498dea539c421e1019c1dd
* This comment will update when the build is complete

### Failed CI Steps
* [Jest Integration Tests](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91399/execution/node/361/log)

### History
* :green_heart: [Build #91326](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91326/) succeeded a56c61e9ff5f3e4172ca589d15a196152c4838ec
* :broken_heart: [Build #91319](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91319/) failed b5f098ffb4cbd1828adb4d4a96a637b2d32f70c0
* :broken_heart: [Build #91290](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91290/) failed 1ae02893c04952a64641ca40a79a30abecdcd77e

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91290/","number":"91290","commit":"1ae02893c04952a64641ca40a79a30abecdcd77e"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91319/","number":"91319","commit":"b5f098ffb4cbd1828adb4d4a96a637b2d32f70c0"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91326/","number":"91326","commit":"a56c61e9ff5f3e4172ca589d15a196152c4838ec"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91399/","number":"91399","commit":"6b6fc49779127fcd80498dea539c421e1019c1dd"}]}
PIPELINE-->

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91491/)
* Commit: a04d0c750955ab8063f13f5f5a982b89bdaf15a5

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Distributable file count
| id | [before](https://github.com/elastic/kibana/commit/67564b9776abd608434c95375968598bfa082849) | [after](https://github.com/elastic/kibana/commit/a04d0c750955ab8063f13f5f5a982b89bdaf15a5) | diff |
| --- | --- | --- | --- |
| `default` | 43182 | 43199 | +17 |
| `oss` | 22579 | 22596 | +17 |

### History
* :broken_heart: [Build #91399](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91399/) failed 6b6fc49779127fcd80498dea539c421e1019c1dd
* :green_heart: [Build #91326](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91326/) succeeded a56c61e9ff5f3e4172ca589d15a196152c4838ec
* :broken_heart: [Build #91319](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91319/) failed b5f098ffb4cbd1828adb4d4a96a637b2d32f70c0
* :broken_heart: [Build #91290](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91290/) failed 1ae02893c04952a64641ca40a79a30abecdcd77e

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91290/","number":"91290","commit":"1ae02893c04952a64641ca40a79a30abecdcd77e"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91319/","number":"91319","commit":"b5f098ffb4cbd1828adb4d4a96a637b2d32f70c0"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91326/","number":"91326","commit":"a56c61e9ff5f3e4172ca589d15a196152c4838ec"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91399/","number":"91399","commit":"6b6fc49779127fcd80498dea539c421e1019c1dd"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91491/","number":"91491","commit":"a04d0c750955ab8063f13f5f5a982b89bdaf15a5"}]}
PIPELINE-->

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91695/)
* Commit: 1077f3049020f511f385ac5770df75a5e78dec4c

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Distributable file count
| id | [before](https://github.com/elastic/kibana/commit/3cb26ebe8d8aaabcf16daf18fe608ab591d2ab77) | [after](https://github.com/elastic/kibana/commit/1077f3049020f511f385ac5770df75a5e78dec4c) | diff |
| --- | --- | --- | --- |
| `default` | 43468 | 43486 | +18 |
| `oss` | 22744 | 22762 | +18 |

### History
* :green_heart: [Build #91491](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91491/) succeeded a04d0c750955ab8063f13f5f5a982b89bdaf15a5
* :broken_heart: [Build #91399](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91399/) failed 6b6fc49779127fcd80498dea539c421e1019c1dd
* :green_heart: [Build #91326](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91326/) succeeded a56c61e9ff5f3e4172ca589d15a196152c4838ec
* :broken_heart: [Build #91319](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91319/) failed b5f098ffb4cbd1828adb4d4a96a637b2d32f70c0
* :broken_heart: [Build #91290](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91290/) failed 1ae02893c04952a64641ca40a79a30abecdcd77e

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91290/","number":"91290","commit":"1ae02893c04952a64641ca40a79a30abecdcd77e"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91319/","number":"91319","commit":"b5f098ffb4cbd1828adb4d4a96a637b2d32f70c0"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91326/","number":"91326","commit":"a56c61e9ff5f3e4172ca589d15a196152c4838ec"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91399/","number":"91399","commit":"6b6fc49779127fcd80498dea539c421e1019c1dd"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91491/","number":"91491","commit":"a04d0c750955ab8063f13f5f5a982b89bdaf15a5"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91695/","number":"91695","commit":"1077f3049020f511f385ac5770df75a5e78dec4c"}]}
PIPELINE-->

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91727/)
* Commit: 8f4581b50ec03221c2b1ececc0629b96576908c5

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Distributable file count
| id | [before](https://github.com/elastic/kibana/commit/3cb26ebe8d8aaabcf16daf18fe608ab591d2ab77) | [after](https://github.com/elastic/kibana/commit/8f4581b50ec03221c2b1ececc0629b96576908c5) | diff |
| --- | --- | --- | --- |
| `default` | 43468 | 43486 | +18 |
| `oss` | 22744 | 22762 | +18 |

### History
* :green_heart: [Build #91695](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91695/) succeeded 1077f3049020f511f385ac5770df75a5e78dec4c
* :green_heart: [Build #91491](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91491/) succeeded a04d0c750955ab8063f13f5f5a982b89bdaf15a5
* :broken_heart: [Build #91399](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91399/) failed 6b6fc49779127fcd80498dea539c421e1019c1dd
* :green_heart: [Build #91326](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91326/) succeeded a56c61e9ff5f3e4172ca589d15a196152c4838ec
* :broken_heart: [Build #91319](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91319/) failed b5f098ffb4cbd1828adb4d4a96a637b2d32f70c0

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91319/","number":"91319","commit":"b5f098ffb4cbd1828adb4d4a96a637b2d32f70c0"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91326/","number":"91326","commit":"a56c61e9ff5f3e4172ca589d15a196152c4838ec"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91399/","number":"91399","commit":"6b6fc49779127fcd80498dea539c421e1019c1dd"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91491/","number":"91491","commit":"a04d0c750955ab8063f13f5f5a982b89bdaf15a5"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91695/","number":"91695","commit":"1077f3049020f511f385ac5770df75a5e78dec4c"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91727/","number":"91727","commit":"8f4581b50ec03221c2b1ececc0629b96576908c5"}]}
PIPELINE-->

pgayvallet commented on src/core/server/logging/integration_tests/logging.test.ts:
```diff
@@ -146,12 +146,18 @@ describe('logging service', () => {
     const loggingConfig$ = new Subject<LoggerContextConfigInput>();
-    const setContextConfig = (enable: boolean) =>
-      enable ? loggingConfig$.next(CUSTOM_LOGGING_CONFIG) : loggingConfig$.next({});
+    const setContextConfig = async (enable: boolean) => {
+      loggingConfig$.next(enable ? CUSTOM_LOGGING_CONFIG : {});
+      // need to wait for config to reload. nextTick is enough, using delay just to be sure
+      await delay(10);
```
This is required because `LoggingService.configure` underlying call to `setContextConfig` (that is now asynchronous) is not awaited.

In real world scenario this is (probably) fine, as configure takes an observable as parameter, but in the tests, we were assuming the provided observable was immediately and synchronously triggering the config reload.

We could change the `LoggingService.setup.configure` implementation to return an observable that would emit when `loggingSystem.setContextConfig` resolves, but that's some additional complexity and didn't felt necessary.

pgayvallet commented on src/core/server/logging/integration_tests/rolling_file_appender.test.ts:
```diff
@@ -0,0 +1,220 @@
+        return delay(nextRolling - now + 1);
+      };
+
+      // wait for a rolling time boundary to minimize the risk to have logs emitted in different intervals
+      // the `1s` interval should be way more than enough to log 2 messages
+      await waitForNextRollingTime();
+
+      logger.info(message(1));
+      logger.info(message(2));
+
+      await waitForNextRollingTime();
```
Integration-testing the `time-interval` policy is a pain, but I really think we do need integration tests for this appender, and I didn't want to mock anything, so I'm stuck with real time-based `await` calls. The test takes ~= 2.5sec to complete, so I think that's alright.

restrry replied:
can we hardcode the delay value to avoid using the internal method `getNextRollingTime`? I'd prefer we consider integration tests as 'black box' testing.

pgayvallet replied:
Using a hardcoded delay would be way more flaky than properly computing the next interval tick. Best I can do would be to use `moment().startOf('second').add('second', 1)` instead of calling `getNextRollingTime`

restrry replied:
nah, let's keep it. `getNextRollingTime` is less cryptic.

pgayvallet commented on src/core/server/logging/logging_system.ts:
```diff
@@ -67,9 +71,12 @@ export class LoggingSystem implements LoggerFactory {
-  public upgrade(rawConfig: LoggingConfigType) {
-    const config = new LoggingConfig(rawConfig)!;
-    this.applyBaseConfig(config);
+  public async upgrade(rawConfig: LoggingConfigType) {
+    // We only want the console appender for the CLI process,
+    // so we use the 'default' configuration as defined by the schema.
+    const usedConfig = this.env.isDevCliParent ? loggingConfig.schema.validate({}) : rawConfig;
+    const config = new LoggingConfig(usedConfig)!;
+    await this.applyBaseConfig(config);
```
During my testing, I added a size-based rolling-file appender to the `root` logger. That caused races issues as both the `CLI` and the `Kibana` processes were trying to roll the file at the same time when the limit was reached.

After discussion with @spalger, we decided that the CLI should not use the logging configuration, and just fallback to the default (which is just outputting to the console)

The multi-process rolling synchronisation problematic should be handled separately from this PR, probably as a part of https://github.com/elastic/kibana/issues/68626.

pgayvallet commented on src/core/server/logging/logging_system.ts:
```diff
@@ -154,17 +161,21 @@ export class LoggingSystem implements LoggerFactory {
     const computedConfig = [...this.contextConfigs.values()].reduce(
       (baseConfig, contextConfig) => baseConfig.extend(contextConfig),
       newBaseConfig
     );
 
+    // reconfigure all the loggers without configuration to have them use the buffer
+    // appender while we are awaiting for the appenders to be disposed.
+    for (const [loggerKey, loggerAdapter] of this.loggers) {
+      loggerAdapter.updateLogger(this.createLogger(loggerKey, undefined));
+    }
+
     // Appenders must be reset, so we first dispose of the current ones, then
     // build up a new set of appenders.
-    for (const appender of this.appenders.values()) {
-      appender.dispose();
-    }
+    await Promise.all([...this.appenders.values()].map((a) => a.dispose()));
```
Small change, heavy implications. 

We were not awaiting for the previous appenders to dispose before creating the new ones (which was a mistake, we are actually doing it anywhere else where we dispose appenders, such as `LoggingSystem.stop`.

Effectively, if this had low to no impact with previous appenders, it could cause race conditions with `rolling-file`, as if a rolling was in progress when going through `applyBaseConfig `, we were not awaiting for it to complete, which causes the newly created appender to also start a rollover, resulting in a race on the files removal/renaming.

If the change is trivial, `applyBaseConfig` is now asynchronous, which required to adapt the consumer chain of this API.

pgayvallet reviewed (commented):
self-review

elasticmachine commented:
Pinging @elastic/kibana-core (Team:Core)

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91795/)
* Commit: a9e21d598fe302a1e76c0b2be21f556321252842

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Distributable file count
| id | [before](https://github.com/elastic/kibana/commit/3cb26ebe8d8aaabcf16daf18fe608ab591d2ab77) | [after](https://github.com/elastic/kibana/commit/a9e21d598fe302a1e76c0b2be21f556321252842) | diff |
| --- | --- | --- | --- |
| `default` | 43468 | 43486 | +18 |
| `oss` | 22744 | 22762 | +18 |

### History
* :green_heart: [Build #91727](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91727/) succeeded 8f4581b50ec03221c2b1ececc0629b96576908c5
* :green_heart: [Build #91695](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91695/) succeeded 1077f3049020f511f385ac5770df75a5e78dec4c
* :green_heart: [Build #91491](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91491/) succeeded a04d0c750955ab8063f13f5f5a982b89bdaf15a5
* :broken_heart: [Build #91399](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91399/) failed 6b6fc49779127fcd80498dea539c421e1019c1dd
* :green_heart: [Build #91326](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91326/) succeeded a56c61e9ff5f3e4172ca589d15a196152c4838ec

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91326/","number":"91326","commit":"a56c61e9ff5f3e4172ca589d15a196152c4838ec"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91399/","number":"91399","commit":"6b6fc49779127fcd80498dea539c421e1019c1dd"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91491/","number":"91491","commit":"a04d0c750955ab8063f13f5f5a982b89bdaf15a5"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91695/","number":"91695","commit":"1077f3049020f511f385ac5770df75a5e78dec4c"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91727/","number":"91727","commit":"8f4581b50ec03221c2b1ececc0629b96576908c5"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91795/","number":"91795","commit":"a9e21d598fe302a1e76c0b2be21f556321252842"}]}
PIPELINE-->

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91806/)
* Commit: 3e740c6ef656fdc7b625f275976d339adeb73c7e

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Distributable file count
| id | [before](https://github.com/elastic/kibana/commit/3cb26ebe8d8aaabcf16daf18fe608ab591d2ab77) | [after](https://github.com/elastic/kibana/commit/3e740c6ef656fdc7b625f275976d339adeb73c7e) | diff |
| --- | --- | --- | --- |
| `default` | 43468 | 43486 | +18 |
| `oss` | 22744 | 22762 | +18 |

### History
* :green_heart: [Build #91795](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91795/) succeeded a9e21d598fe302a1e76c0b2be21f556321252842
* :green_heart: [Build #91727](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91727/) succeeded 8f4581b50ec03221c2b1ececc0629b96576908c5
* :green_heart: [Build #91695](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91695/) succeeded 1077f3049020f511f385ac5770df75a5e78dec4c
* :green_heart: [Build #91491](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91491/) succeeded a04d0c750955ab8063f13f5f5a982b89bdaf15a5
* :broken_heart: [Build #91399](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91399/) failed 6b6fc49779127fcd80498dea539c421e1019c1dd

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91399/","number":"91399","commit":"6b6fc49779127fcd80498dea539c421e1019c1dd"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91491/","number":"91491","commit":"a04d0c750955ab8063f13f5f5a982b89bdaf15a5"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91695/","number":"91695","commit":"1077f3049020f511f385ac5770df75a5e78dec4c"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91727/","number":"91727","commit":"8f4581b50ec03221c2b1ececc0629b96576908c5"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91795/","number":"91795","commit":"a9e21d598fe302a1e76c0b2be21f556321252842"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91806/","number":"91806","commit":"3e740c6ef656fdc7b625f275976d339adeb73c7e"}]}
PIPELINE-->

legrego commented on src/core/server/logging/appenders/rolling_file/strategies/index.ts:
```diff
@@ -0,0 +1,39 @@
+export { RollingStrategy } from './strategy';
+export type RollingStrategyConfig = NumericRollingStrategyConfig;
+
+export const rollingStrategyConfigSchema = schema.oneOf([numericRollingStrategyConfigSchema]);
```
Since we only have a single strategy available, what do you think about providing a default value, so that consumers aren't forced to specify it? I mistakenly omitted this from my yml, and the `@kbn/config-schema` error message was hard to decipher at best.

pgayvallet replied:
Yea, I even added default values for both the policy and the strategy at some point, so that

```yaml
rolling-file:
    kind: rolling-file
    path: /Users/pierregayvallet/Documents/test-log/kibana.log
    layout:
        kind: pattern
```

would be enough, then I reverted it, thinking that we didn't do that for other appenders (like you must always specify `layout`, there isn't a default value for it).

@elastic/kibana-core do you have any opinion on this?

legrego commented on src/core/server/logging/appenders/rolling_file/strategies/numeric/pattern_matcher.ts:
```diff
@@ -0,0 +1,81 @@
+const escape = (string: string) => string.replace(/[.*+\-?^${}()|[\]\\]/g, '\\$&');
+
+const createNumericMatcher = (fileBaseName: string, pattern: string): RegExp => {
+  let extStart = fileBaseName.indexOf('.');
```
Do you want `lastIndexOf` here instead? Would this incorrectly say that `foo.bar.log` has an extension of `bar.log`?

```suggestion
  let extStart = fileBaseName.lastIndexOf('.');
```

pgayvallet replied:
That was intentional. `foo.bar.log` should (imho) roll to `foo-1.bar.log`? That's usually what is done for multi-extension I think?

legrego replied:
Ah, I think I've personally seen `foo.bar-1.log` in a past life, but I'm fine to defer to you if this was intentional. I was thinking if someone had `kibana.audit.json`, that `kibana-1.audit.json` would look a bit funny.

pgayvallet replied:
I was think of `foo.tar.gz`, where the rename would make more sense to be `foo-1.tar.gz`. But I guess it depends then. We could eventually add an option later. For now, I good with either option.

legrego replied:
> For now, I good with either option.

Same. Feel free to leave it as-is

legrego commented on src/core/server/logging/appenders/rolling_file/strategies/numeric/pattern_matcher.ts:
```diff
@@ -0,0 +1,81 @@
+ * under the License.
+ */
+
+const escape = (string: string) => string.replace(/[.*+\-?^${}()|[\]\\]/g, '\\$&');
```
**optional nit** I know that lodash is sometimes overused, but they offer [`_.escapeRegExp()`](https://lodash.com/docs/4.17.15#escapeRegExp) so that you don't have to re-invent or re-test this.

legrego commented on src/core/server/logging/appenders/rolling_file/strategies/numeric/pattern_matcher.ts:
```diff
@@ -0,0 +1,81 @@
+  pattern: string,
+  index: number
+): string => {
+  let suffixStart = fileBaseName.indexOf('.');
```
Same question here w/r/t `lastIndexOf`

pgayvallet replied:
resolving for readability, will be included in https://github.com/elastic/kibana/pull/84735#discussion_r536121024

legrego commented on src/core/server/logging/appenders/rolling_file/strategies/numeric/rolling_tasks.ts:
```diff
@@ -0,0 +1,96 @@
+  filesToDelete: string[];
+}) => {
+  for (const fileToDelete of filesToDelete) {
+    await unlink(join(logFileFolder, fileToDelete));
```
Do we need any exception handling/logging for this?

pgayvallet replied:
Some tasks are tricky to resume in case of atomic errors. For instance when rolling files in order, if there is an error when renaming a file, we can't really roll the next one as it would throw because the rename target still exists. This is why I'm globally catching all errors from the rolling strategy directly in the RollingFileHandler instead. Do you think this could be improved?

Now for logging the error, I overall agree, see https://github.com/elastic/kibana/pull/84735#discussion_r535490305

legrego replied:
I agree that resumption is tricky/impossible in some of these scenarios. I can't think of a way to improve what we have, as long as the errors that we're logging are descriptive enough by the time they reach the `RollingFileHandler`. Filesystem errors are usually descriptive enough(?), so I think what you have is fine.

I was asking a legitimate question here, I wasn't necessarily flagging this as an issue

legrego commented on src/core/server/logging/appenders/rolling_file/strategies/numeric/rolling_tasks.ts:
```diff
@@ -0,0 +1,96 @@
+  for (let i = filesToRoll.length - 1; i >= 0; i--) {
+    const oldFileName = filesToRoll[i];
+    const newFileName = getRollingFileName(logFileBaseName, pattern, i + 2);
+    await rename(join(logFileFolder, oldFileName), join(logFileFolder, newFileName));
```
Same question here w/r/t exception handling

legrego reviewed (commented):
I tested this new appender for Kibana's new audit log, and it's working very nicely! This is a really impressive PR, thanks again for putting this up so quickly

pgayvallet commented on src/core/server/logging/appenders/rolling_file/rolling_file_context.ts:
```diff
@@ -0,0 +1,46 @@
+      this.currentFileTime = birthtime.getTime();
+      this.currentFileSize = size;
+    } catch (e) {
+      this.currentFileTime = Date.now();
```
Ideally, we would only catch `ENOENT` (or such) errors. But that's kinda the first time we got a logger/appender implementation complex enough to potentially throw errors, so I'm not really sure how we should handle failures internal to the logger system or its appenders, so for safety I used a catch-all instead.

pgayvallet replied:
Good point, will add a `ENOENT` check and logs the error.

Remaining question (same as in https://github.com/elastic/kibana/pull/84735#discussion_r535490305) is, should we:
1. use `console.log`
2. use `console.error`
3. create a `ConsoleAppender` in the `RollingFileAppender` and use it internally for all error logging (we would use the same `layout` as the one provided to the rolling file appender). In that case, it will logs in `stdout` as if we were using `console.log`

wdyt?

restrry replied:
The 3rd option looks fine. However, I'd expect that it uses the default `layout` that a user specifies for the `root` context.
If the 3rd option is too hard to implement, the 2nd one is well enough.

pgayvallet commented on src/core/server/logging/appenders/rolling_file/rolling_file_appender.ts:
```diff
@@ -0,0 +1,174 @@
+      return;
+    }
+    this.isRolling = true;
+    try {
+      await this.strategy.rollout();
+      await this.fileManager.closeStream();
+    } catch (e) {
+      // eslint-disable-next-line no-console
+      console.log('Error while rolling file: ', e);
+    }
```
> Would a failed rollout never, sometimes, or always prevent future logs from being written?

it's never `never` when you interact with filesystem, but except in case of hard I/O failure, I think the worth case would be that we would not be able to roll the log file, and that the next rolling cycle would still write in the previous (non-renamed) file. Which seemed acceptable (also we can't really do anything except warning about the failure)

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/92267/)
* Commit: 6694e24a1cf6bff9d3f2f467cbe7f220a7f2cc7e

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Distributable file count
| id | [before](https://github.com/elastic/kibana/commit/2c7a85dd38f025000f27d3c29bb7a625f0227a1d) | [after](https://github.com/elastic/kibana/commit/6694e24a1cf6bff9d3f2f467cbe7f220a7f2cc7e) | diff |
| --- | --- | --- | --- |
| `default` | 46897 | 47675 | +778 |
| `oss` | 27565 | 27583 | +18 |

### History
* :green_heart: [Build #91806](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91806/) succeeded 3e740c6ef656fdc7b625f275976d339adeb73c7e
* :green_heart: [Build #91795](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91795/) succeeded a9e21d598fe302a1e76c0b2be21f556321252842
* :green_heart: [Build #91727](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91727/) succeeded 8f4581b50ec03221c2b1ececc0629b96576908c5
* :green_heart: [Build #91695](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91695/) succeeded 1077f3049020f511f385ac5770df75a5e78dec4c
* :green_heart: [Build #91491](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91491/) succeeded a04d0c750955ab8063f13f5f5a982b89bdaf15a5

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91491/","number":"91491","commit":"a04d0c750955ab8063f13f5f5a982b89bdaf15a5"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91695/","number":"91695","commit":"1077f3049020f511f385ac5770df75a5e78dec4c"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91727/","number":"91727","commit":"8f4581b50ec03221c2b1ececc0629b96576908c5"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91795/","number":"91795","commit":"a9e21d598fe302a1e76c0b2be21f556321252842"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91806/","number":"91806","commit":"3e740c6ef656fdc7b625f275976d339adeb73c7e"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/92267/","number":"92267","commit":"6694e24a1cf6bff9d3f2f467cbe7f220a7f2cc7e"}]}
PIPELINE-->

restrry commented on src/core/server/logging/appenders/rolling_file/strategies/numeric/rolling_tasks.ts:
```diff
@@ -0,0 +1,96 @@
+  logFileFolder: string;
+  filesToDelete: string[];
+}) => {
+  for (const fileToDelete of filesToDelete) {
```
can we run deletion operation in parallel?

restrry commented on src/core/server/logging/appenders/rolling_file/strategies/numeric/rolling_tasks.test.ts:
```diff
@@ -0,0 +1,166 @@
+
+  describe('getOrderedRolledFiles', () => {
+    it('returns the rolled files matching the pattern in order', async () => {
+      readdirMock.mockResolvedValue(['kibana-3.log', 'kibana-1.log', 'kibana-2.log']);
```
Could you add a test that it handles sorting order correctly?
```js
readdirMock.mockResolvedValue(['kibana-10.log', 'kibana-1.log', 'kibana-11.log']);
```

restrry commented on src/core/server/logging/appenders/rolling_file/strategies/numeric/numeric_strategy.ts:
```diff
@@ -0,0 +1,149 @@
+      }
+    },
+  }),
+  max: schema.number({ min: 1, defaultValue: 7 }),
```
should have max limit, I suppose

restrry commented on src/core/server/logging/appenders/rolling_file/strategies/numeric/numeric_strategy.ts:
```diff
@@ -0,0 +1,149 @@
+    await rollCurrentFile({ pattern, logFileBaseName, logFileFolder });
+
+    // updates the context file info to mirror the new size and date
+    this.context.refreshFileInfo();
```
could you add a comment that it's necessary for the time based policy

restrry commented on src/core/server/logging/integration_tests/rolling_file_appender.test.ts:
```diff
@@ -0,0 +1,220 @@
+    readFileSync(join(testDir, basename)).toString('utf-8');
+
+  beforeEach(async () => {
+    testDir = mkdtempSync('rolling-test');
```
why not async counterparts?

pgayvallet replied:
No reasons, I guess it was just late when I did it. Will change

restrry commented on src/core/server/logging/integration_tests/rolling_file_appender.test.ts:
```diff
@@ -0,0 +1,220 @@
+
+      const logger = root.logger.get('test.rolling.file');
+
+      // size = 100b, message.length ~= 40b, should roll every 3 message
```
we create a buffer of fixed size to make sure to have better control over the rolling process 
```js
const asciiBuf = Buffer.alloc(5, 'a', 'utf8');
```

pgayvallet replied:
I'm only using ascii chars in the message, so we kinda know that `sizeInBytes === msg.length`, but I can use a buffer to add more control if we want to.

restrry commented on src/core/server/logging/integration_tests/rolling_file_appender.test.ts:
```diff
@@ -0,0 +1,220 @@
+        },
+        policy: {
+          kind: 'time-interval',
+          interval: '1s',
```
that might be unstable on CI 😞

restrry commented on src/core/server/logging/logging_system.ts:
```diff
@@ -67,9 +71,12 @@ export class LoggingSystem implements LoggerFactory {
+  public async upgrade(rawConfig: LoggingConfigType) {
+    // We only want the console appender for the CLI process,
+    // so we use the 'default' configuration as defined by the schema.
+    const usedConfig = this.env.isDevCliParent ? loggingConfig.schema.validate({}) : rawConfig;
```
can we keep this logic outside of logging system? we ca make `rawConfig?: LoggingConfigType` to make sure it falls back to the default config

pgayvallet commented on src/core/server/logging/appenders/rolling_file/policies/size_limit/size_limit_policy.ts:
```diff
@@ -0,0 +1,58 @@
+
+export const sizeLimitTriggeringPolicyConfigSchema = schema.object({
+  kind: schema.literal('size-limit'),
+  size: schema.byteSize({ min: '1b', defaultValue: '100mb' }),
```
I haven't. @Kushmaro maybe you can help us here?

Kushmaro replied:
By our [log delivery feature design](https://docs.google.com/document/d/1mbZ5zKlX9-6OSn4DmTX_RthUg_na4LFUsVZ2-cQPz0k/edit?usp=sharing#heading=h.lwfghv6kn3o4) (see storage section) we concluded that the smallest kibana instance in cloud will have around 1GB of storage space

is the 100mb referring to the log size rotation limit? cloud also enforces a quota on the log path, and I * think * that quota is 100mb total, so maybe a more sensible rotation limit would be 10mb.

@s-nel for awareness and further guidelines (he's on PTO though so may take a while)

pgayvallet replied:
> is the 100mb referring to the log size rotation limit

That's the maximum size of the log file before it got rolled out.

s-nel replied:
We use 100MB rollover size for ES audit logs and keep 2 files. We might choose something smaller for Kibana since the instances don't have much storage.

pgayvallet commented on src/core/server/logging/appenders/rolling_file/policies/time_interval/time_interval_policy.ts:
```diff
@@ -0,0 +1,96 @@
+    defaultValue: '24h',
+    validate: (interval) => {
+      if (!isValidRolloverInterval(interval)) {
+        return 'Interval value cannot overflow to a higher field.';
```
The message may be misleading. What we can't have it a non-integer value for the period, because the underlying logic for fixed boundaries rolling does not allow it. We can have `60mins` (1h), but not `90mins` (1,5h).

pgayvallet commented on src/core/server/logging/appenders/rolling_file/rolling_file_appender.ts:
```diff
@@ -0,0 +1,174 @@
+   * Formats specified `record` and writes it to the specified file. If the record
+   * would trigger a rollover, it will be performed before the effective write operation.
+   */
+  public append(record: LogRecord) {
```
Not sure to understand,

Internally, once disposed, we will never call `append`. Even if some messages where queued into the buffer, we'll directly write them to file (`_writeToFile`) instead of calling `append` (to avoid the risk of triggering another rolling when the appender is disposed).

Externally, nothing is currently forbidding to call `append` once the appender is disposed, but that's something already present in our other appenders (you can call `append` on `FileAppender` once disposed), it's the logging system responsibility to avoid such scenarios.

restrry replied:
>  but that's something already present in our other appenders 

it wasn't the problem because `dispose` used to be synchronous. Ok, it shouldn't block `RollingFile` appender merge

pgayvallet commented on src/core/server/logging/appenders/rolling_file/rolling_file_appender.test.ts:
```diff
@@ -0,0 +1,275 @@
+  });
+
+  afterAll(() => {
+    resetAllMocks();
```
I'm mocking some class constructors in the test file, so yea.

```ts
// src/core/server/logging/appenders/rolling_file/rolling_file_appender.test.mocks.ts#L42
export const RollingFileManagerMock = jest.fn();
jest.doMock('./rolling_file_manager', () => ({
  RollingFileManager: RollingFileManagerMock,
}));
```

```ts
// src/core/server/logging/appenders/rolling_file/rolling_file_appender.test.ts#L101
context = rollingFileAppenderMocks.createContext('file-path');
RollingFileContextMock.mockImplementation(() => context);
```

I could just `clear` them instead of `reset` as the implementations are redefined during the `beforeEach` block, but as the call was required anyway, `reset` felt better.

restrry replied:
I'm not sure how jest runs tests. But if it runs them in parallel, the shared state could cause false-positive test failures.

pgayvallet commented on src/core/server/logging/README.md:
```diff
@@ -127,6 +131,138 @@ Outputs the process ID.
+      kind: rolling-file
+      path: /var/logs/kibana.log
+      policy:
+        kind: size-limit
```
Well, IDK about doing it directly in this PR instead of globalizing all the required changes in https://github.com/elastic/kibana/issues/57551 tbh. We would be introducing inconsistencies in our own logging API, which is imho probably worse.

In example:

> size-limit --> SizeBasedTriggeringPolicy

All our other 'types' are lowercase, dash separated. Having a mix of both seems wrong

> policy --> policies

In log4j configuration , `policies` is a map (they accept multiple policies). we don't ATM (and TIL by looking at log4j's code, it would requires significant work, their code is a patchwork of hacks and side effects to support that...) I feel like using `policies` could be misleading as the user may try to use an array instead of an object in the configuration?

Also, note that if we want to align the naming, there are other things to change:

`kind: size-limit`, in log4j, it's `type`, not `kind`. However all our other logging configuration structures are using the `kind` naming, so this would introduce yet more divergence in our own logging config API.

Not opposed to do it right now if you think it's still better though, just tell me.

restrry replied:
>  in log4j, it's type, not kind

yeah, https://github.com/elastic/kibana/issues/57551 should fix it. 

> All our other 'types' are lowercase, dash separated.

Just curious, what `types` do you mean? I'm fine if we postpone this change till https://github.com/elastic/kibana/issues/57551 also it means that we should do that before removing the legacy log rotation @joshdover any objections?

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/92791/)
* Commit: 3aa0bc700443fa2323b61e4f5955cc3974863505

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Distributable file count
| id | [before](https://github.com/elastic/kibana/commit/e4b89d42d8bd16d92f94cff6ade7abd221ea0594) | [after](https://github.com/elastic/kibana/commit/3aa0bc700443fa2323b61e4f5955cc3974863505) | diff |
| --- | --- | --- | --- |
| `default` | 46947 | 47724 | +777 |
| `oss` | 27573 | 27590 | +17 |

### History
* :green_heart: [Build #92267](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/92267/) succeeded 6694e24a1cf6bff9d3f2f467cbe7f220a7f2cc7e
* :green_heart: [Build #91806](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91806/) succeeded 3e740c6ef656fdc7b625f275976d339adeb73c7e
* :green_heart: [Build #91795](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91795/) succeeded a9e21d598fe302a1e76c0b2be21f556321252842
* :green_heart: [Build #91727](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91727/) succeeded 8f4581b50ec03221c2b1ececc0629b96576908c5
* :green_heart: [Build #91695](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91695/) succeeded 1077f3049020f511f385ac5770df75a5e78dec4c

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91695/","number":"91695","commit":"1077f3049020f511f385ac5770df75a5e78dec4c"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91727/","number":"91727","commit":"8f4581b50ec03221c2b1ececc0629b96576908c5"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91795/","number":"91795","commit":"a9e21d598fe302a1e76c0b2be21f556321252842"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91806/","number":"91806","commit":"3e740c6ef656fdc7b625f275976d339adeb73c7e"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/92267/","number":"92267","commit":"6694e24a1cf6bff9d3f2f467cbe7f220a7f2cc7e"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/92791/","number":"92791","commit":"3aa0bc700443fa2323b61e4f5955cc3974863505"}]}
PIPELINE-->

pgayvallet commented:
@restrry I addressed most of your comments. The points I still need input on are
- https://github.com/elastic/kibana/pull/84735#discussion_r539065609
- https://github.com/elastic/kibana/pull/84735#discussion_r538287432
- https://github.com/elastic/kibana/pull/84735#discussion_r538545314 

PTAL

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/93004/)
* Commit: bf0e390fd79a02ac6e4500b69b741bcb1801bfd8

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Distributable file count
| id | [before](https://github.com/elastic/kibana/commit/f2d961df6a5f78807dc2ef265aaba3c96d691881) | [after](https://github.com/elastic/kibana/commit/bf0e390fd79a02ac6e4500b69b741bcb1801bfd8) | diff |
| --- | --- | --- | --- |
| `default` | 47739 | 47756 | +17 |
| `oss` | 27594 | 27611 | +17 |

### History
* :green_heart: [Build #92791](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/92791/) succeeded 3aa0bc700443fa2323b61e4f5955cc3974863505
* :green_heart: [Build #92267](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/92267/) succeeded 6694e24a1cf6bff9d3f2f467cbe7f220a7f2cc7e
* :green_heart: [Build #91806](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91806/) succeeded 3e740c6ef656fdc7b625f275976d339adeb73c7e
* :green_heart: [Build #91795](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91795/) succeeded a9e21d598fe302a1e76c0b2be21f556321252842
* :green_heart: [Build #91727](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91727/) succeeded 8f4581b50ec03221c2b1ececc0629b96576908c5

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91727/","number":"91727","commit":"8f4581b50ec03221c2b1ececc0629b96576908c5"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91795/","number":"91795","commit":"a9e21d598fe302a1e76c0b2be21f556321252842"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91806/","number":"91806","commit":"3e740c6ef656fdc7b625f275976d339adeb73c7e"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/92267/","number":"92267","commit":"6694e24a1cf6bff9d3f2f467cbe7f220a7f2cc7e"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/92791/","number":"92791","commit":"3aa0bc700443fa2323b61e4f5955cc3974863505"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/93004/","number":"93004","commit":"bf0e390fd79a02ac6e4500b69b741bcb1801bfd8"}]}
PIPELINE-->

restrry reviewed (approved):
We can change `size-limit` policy defaults in a follow up

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/93379/)
* Commit: bbb96f4d2b45282b9b72dfe96882d98082f6a30f

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Distributable file count
| id | [before](https://github.com/elastic/kibana/commit/0f408041b48a03fa46492f62d08fe5d8025390ea) | [after](https://github.com/elastic/kibana/commit/bbb96f4d2b45282b9b72dfe96882d98082f6a30f) | diff |
| --- | --- | --- | --- |
| `default` | 46991 | 47768 | +777 |
| `oss` | 27598 | 27615 | +17 |

### History
* :green_heart: [Build #93004](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/93004/) succeeded bf0e390fd79a02ac6e4500b69b741bcb1801bfd8
* :green_heart: [Build #92791](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/92791/) succeeded 3aa0bc700443fa2323b61e4f5955cc3974863505
* :green_heart: [Build #92267](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/92267/) succeeded 6694e24a1cf6bff9d3f2f467cbe7f220a7f2cc7e
* :green_heart: [Build #91806](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91806/) succeeded 3e740c6ef656fdc7b625f275976d339adeb73c7e
* :green_heart: [Build #91795](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91795/) succeeded a9e21d598fe302a1e76c0b2be21f556321252842

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91795/","number":"91795","commit":"a9e21d598fe302a1e76c0b2be21f556321252842"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/91806/","number":"91806","commit":"3e740c6ef656fdc7b625f275976d339adeb73c7e"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/92267/","number":"92267","commit":"6694e24a1cf6bff9d3f2f467cbe7f220a7f2cc7e"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/92791/","number":"92791","commit":"3aa0bc700443fa2323b61e4f5955cc3974863505"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/93004/","number":"93004","commit":"bf0e390fd79a02ac6e4500b69b741bcb1801bfd8"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/93379/","number":"93379","commit":"bbb96f4d2b45282b9b72dfe96882d98082f6a30f"}]}
PIPELINE-->
